pkg/monitor/cluster/healthz.go

```diff
@@ -5,17 +5,53 @@ package cluster
 
 import (
 	"strconv"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/errors"
 )
 
 func (mon *Monitor) emitAPIServerHealthzCode() (int, error) {
 	var statusCode int
-	err := mon.cli.Discovery().RESTClient().
+	raw, err := mon.cli.Discovery().RESTClient().
 		Get().
-		AbsPath("/healthz").
+		AbsPath("/readyz").
 		Do().
 		StatusCode(&statusCode).
-		Error()
+		Raw()
+
+	// HTTP 500s are expected behaviour (e.g. on startup/shutdown of the API server) and should not cause this function to return an error. Therefore, silence it.
+	if err != nil && errors.IsUnexpectedServerError(err) {
+		err = nil
+	}
+
+	if statusCode == 200 && err == nil {
+		// apiserver is ready
+		mon.emitGauge("apiserver.ready", 1, nil)
+	} else if statusCode != 200 && err == nil {
+		// apiserver is not ready, but is still reporting status
+		var emitted bool
+		failures := strings.Split(string(raw), "\n")
+
+		for _, failure := range failures {
+			if strings.HasPrefix(failure, "[-]") {
+				failedUnit := failure[3:strings.Index(failure, " ")]
+				mon.emitGauge("apiserver.ready", 0, map[string]string{
+					"failedUnit": failedUnit,
+				})
+				emitted = true
+			}
+		}
+
+		if !emitted {
+			// If we can't emit any failed units in particular, just emit a blank one
+			mon.emitGauge("apiserver.ready", 0, nil)
+		}
+	} else {
+		// apiserver cannot be reached
+		mon.emitGauge("apiserver.ready", 0, nil)
+	}
 
+	// compat?
```

Comment:
So the behaviour below has actually been frustrating me, as someone who's been trying to improve our alerting based on this metric, so let me comment on that briefly...

Right now, the `apiserver.healthz.code` metric emits one timeseries per healthz code (i.e. I get one metric for 500s, one for 200s, one for no response modeled as 0s, etc.) which is difficult to aggregate.

I much prefer the model of "up/down" with 0/1 values here, tbh, and I think it's a good idea to keep both metrics for compatibility given that we have a significant pile of alerts/dashboards built against it. I agree with @m1kola that it's a bit confusing to poll a different endpoint for this but not update the name; we might want to add the new metric totally separately.

Reply:
WRT the different name -- that's reasonable, adding it separately. `healthz` and `readyz` are identical, apart from that `readyz` starts returning failure during the shutdown grace period from an added "shutdown" check, while `healthz` does not. `livez` has different semantics, though. It seems (from the comment at https://github.com/kubernetes/kubernetes/blob/88c1386e0cd39e1939a923e44e799bc639ee4275/staging/src/k8s.io/apiserver/pkg/server/healthz.go#L37-L42 ) that `readyz` and `livez` are meant to replace `healthz`.

the shutdown check: 
https://github.com/kubernetes/kubernetes/blob/88c1386e0cd39e1939a923e44e799bc639ee4275/staging/src/k8s.io/apiserver/pkg/server/healthz.go#L84-L89 
https://github.com/kubernetes/kubernetes/blob/88c1386e0cd39e1939a923e44e799bc639ee4275/staging/src/k8s.io/apiserver/pkg/server/healthz.go#L115-L132